src/ElfUtils/ElfFile.cpp

```diff
@@ -134,6 +138,42 @@ ElfFileImpl<ElfT>::ElfFileImpl(std::filesystem::path file_path,
   InitSections();
 }
 
+template <typename ElfT>
+void ElfFileImpl<ElfT>::InitDynamicEntries(const llvm::object::ELFFile<ElfT>* elf_file) {
+  std::optional<uint64_t> soname_offset;
+  std::optional<llvm::StringRef> dynamic_string_table;
+  auto dynamic_entries_or_error = elf_file->dynamicEntries();
+  if (!dynamic_entries_or_error) {
+    LOG("Unable to get dynamic entries from \"%s\": %s", file_path_.string(),
+        llvm::toString(dynamic_entries_or_error.takeError()));
+    return;
+  }
+
+  auto dyn_range = dynamic_entries_or_error.get();
+  for (const auto& dyn_entry : dyn_range) {
+    switch (dyn_entry.getTag()) {
+      case llvm::ELF::DT_SONAME:
+        soname_offset.emplace(dyn_entry.getVal());
+        break;
+      case llvm::ELF::DT_STRTAB: {
+        auto strtab_addr_or_error = elf_file->toMappedAddr(dyn_entry.getPtr());
+        if (!strtab_addr_or_error) {
+          LOG("Unable to get dynamic string table from DT_STRTAB: %s",
+              llvm::toString(strtab_addr_or_error.takeError()));
+          break;
+        }
+        dynamic_string_table.emplace(absl::bit_cast<const char*>(strtab_addr_or_error.get()));
+        break;
+      }
+      default:
+        break;
+    }
+  }
+  if (soname_offset.has_value() && dynamic_string_table.has_value()) {
+    soname_ = dynamic_string_table.value().data() + soname_offset.value();
```

Comment:
I assume this works because `dynamic_string_table` is a null-separated list of strings?
What happens if `SONAME` is the laste entry in the list? Does LLVM ensure that `dynamic_string_table` ends with a null-character or should we also take the length into account?

Reply:
come to think - we should probably never trust that format is correct and consistent, I am adding boundary checks and null-termination check.

Reply:
done.